docs/proposals/locking.md

```diff
@@ -108,3 +108,198 @@ support:
+        // time that the lock stopped being active. If the lock is still active,
+        // the server can either a) not send this field, or b) send the
+        // zero-value of time.Time.
+        ExpiresAt time.Time `json:"expired_at,omitempty"`
+}
+
+// Active returns whether or not the given lock is still active against the file
+// that it is protecting.
+func (l *Lock) Active() bool {
+        return time.IsZero(l.ExpiresAt)
+}
+```
+
+### Proposed Commands
+
+#### `git lfs lock <path>`
+
+The `lock` command will be used in accordance with the multi-branch flow as
+proposed above to request that lock be granted to the specific path passed an
+argument to the command.
+
+```go
+// LockRequest encapsulates the payload sent across the API when a client would
+// like to obtain a lock against a particular path on a given remote.
+type LockRequest struct {
+        // Path is the path that the client would like to obtain a lock against.
+        Path      string `json:"path"`
+        // Remote is the remote on which the client would like to obtain the
+        // lock.
+        Remote    string `json:"remote"`
+        // LatestRemoteCommit is the SHA of the last known commit from the
+        // remote that we are trying to create the lock against, as found in
+        // `.git/refs/origin/<name>`.
+        LatestRemoteCommit string `json:"latest_remote_commit"`
+        // Committer is the individual that wishes to obtain the lock.
+        Committer struct {
+              // Name is the name of the individual who would like to obtain the
+              // lock, for instance: "Rick Olson".
+              Name string `json:"name"`
+              // Email is the email assopsicated with the individual who would
+              // like to obtain the lock, for instance: "rick@github.com".
+              Email string `json:"email"`
+        } `json:"committer"`
+}
+```
+
+```go
+// LockResponse encapsulates the information sent over the API in response to
+// a `LockRequest`.
+type LockResponse struct {
+        // Lock is the Lock that was optionally created in response to the
+        // payload that was sent (see above). If the lock already exists, then
+        // the existing lock is sent in this field instead, and the author of
+        // that lock remains the same, meaning that the client failed to obtain
+        // that lock. An HTTP status of "409 - Conflict" is used here.
+        //
+        // If the lock was unable to be created, this field will hold the
+        // zero-value of Lock and the Err field will provide a more detailed set
+        // of information.
+        //
+        // If an error was experienced in creating this lock, then the
+        // zero-value of Lock should be sent here instead.
+        Lock Lock `json:"lock"`
+        // CommitNeeded holds the minimum commit SHA that client must have to
+        // obtain the lock.
+        CommitNeeded string `json:"commit_needed"`
```

Comment:
Presumably `CommitNeeded` will only be populated if there is an error (and therefore no lock)? Implication being the client needs to checkout that commit or cherry pick the state, and also may have to `git fetch` if it doesn't have that commit locally.